### DISCUSSION TITLE
Why does the prove time does not change based on the input?

### DISCUSSION THREAD

**Discussion by 2pir2** [Consensys/gnark#1287] [Q&A]
I am trying to see the time it takes for the mimc function on different size of inputs string, However, it seems that the time is not sensitive to the size of the inputs, where I am a bit confused about. I have try inputs of string length 1 and string length 120,000, and the time it takes is pretty similar.
Here is my code:

package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/big"
	"os"
	"time"

	"github.com/consensys/gnark-crypto/ecc"
	"github.com/consensys/gnark/backend/groth16"
	"github.com/consensys/gnark/frontend"
	r1cs2 "github.com/consensys/gnark/frontend/cs/r1cs"
	"github.com/consensys/gnark/std/hash/mimc"
)

const (
	pubInputFile = "public.json"
	priInputFile = "private.json"
	vkKeyFile    = "vk.g16vk"
	proofFile    = "proof.g16p"
)

// Structure to hold the input data from the JSON file
type Inputs struct {
	PreImage string `json:"PreImage"`
	Hash     string `json:"Hash"`
}

type Circuit struct {
	PreImage frontend.Variable
	Hash     frontend.Variable `gnark:",public"`
}

func (circuit *Circuit) Define(api frontend.API) error {
	// Create a new MiMC hash instance
	mimcHash, _ := mimc.NewMiMC(api)

	// Hash the PreImage and store it in a variable
	mimcHash.Write(circuit.PreImage)
	computedHash := mimcHash.Sum()

	// Assert that the computed hash equals the given public hash
	api.AssertIsEqual(circuit.Hash, computedHash)
	return nil
}

func main() {
	// Step 1: Read input data from JSON file
	inputFile := "inputs.json"
	file, err := os.Open(inputFile)
	if err != nil {
		fmt.Printf("Failed to open the file: %v\n", err)
		return
	}
	defer file.Close()

	byteValue, _ := ioutil.ReadAll(file)

	var inputs Inputs
	err = json.Unmarshal(byteValue, &inputs)
	if err != nil {
		fmt.Printf("Failed to parse the JSON file: %v\n", err)
		return
	}

	// Step 2: Convert PreImage and Hash from the JSON to big.Int
	preImage := new(big.Int)
	preImage.SetString(inputs.PreImage, 10) // Base 10 for parsing

	hash := new(big.Int)
	hash.SetString(inputs.Hash, 10) // Base 10 for parsing

	// Step 3: Print PreImage and Hash

	var circuit Circuit

	// Step 4: Compile the circuit

	// Step 6: Creating the witness (assignment) for the circuit
	assignment := &Circuit{
		PreImage: frontend.Variable(preImage),
		Hash:     frontend.Variable(hash), // Directly pass the *big.Int hash as a Variable
	}
	fmt.Println("Expected Hash", assignment.PreImage)
	start := time.Now()
	r1cs, err := frontend.Compile(ecc.BN254.ScalarField(), r1cs2.NewBuilder, &circuit)
	if err != nil {
		fmt.Printf("Compile failed: %v\n", err)
		return
	}

	// Step 5: Groth16 Setup
	pk, vk, err := groth16.Setup(r1cs)
	if err != nil {
		fmt.Printf("Setup failed: %v\n", err)
		return
	}
	// Step 7: Generate the witness for proof generation
	witness, err := frontend.NewWitness(assignment, ecc.BN254.ScalarField())
	if err != nil {
		fmt.Printf("Witness generation failed: %v\n", err)
		return
	}

	// Step 8: Proving the circuit

	proof, err := groth16.Prove(r1cs, pk, witness)
	if err != nil {
		fmt.Printf("Prove failed: %v\n", err)
		return
	}
	elapsed := time.Since(start)
	fmt.Println("Time for Proof:", elapsed)
	// Step 9: Creating the public witness for verification
	publicAssignment := &Circuit{
		Hash: frontend.Variable(hash), // Directly pass the *big.Int hash for public verification
	}

	// Write the verification key to a file
	vkF, err := os.Create(vkKeyFile)
	if err != nil {
		fmt.Errorf("failed to create %s: %v", vkKeyFile, err)
	}
	defer vkF.Close()

	_, err = vk.WriteTo(vkF)
	if err != nil {
		fmt.Errorf("failed to write verification key: %v", err)
	}

	// Write the proof to a file
	proofF, err := os.Create(proofFile)
	if err != nil {
		fmt.Errorf("failed to create %s: %v", proofFile, err)
	}
	defer proofF.Close()

	_, err = proof.WriteTo(proofF)
	if err != nil {
		fmt.Errorf("failed to write proof: %v", err)
	}

	publicWitness, err := frontend.NewWitness(publicAssignment, ecc.BN254.ScalarField(), frontend.PublicOnly())
	if err != nil {
		fmt.Printf("Public witness generation failed: %v\n", err)
		return
	}

	// Step 10: Verifying the proof
	err = groth16.Verify(proof, vk, publicWitness)
	if err != nil {
		fmt.Printf("Verification failed: %v\n", err)
		return
	}
	fmt.Printf("Verification succeeded\n")
}

**Comment by ivokub** [comment]
You are measuring the circuit compile, Groth16 and Groth16 proving. It would be more suitable to only measure the proving time i.e. the part 
```go
proof, err := groth16.Prove(r1cs, pk, witness)
if err != nil {
	fmt.Printf("Prove failed: %v\n", err)
	return
}
```

**Reply by 2pir2** [reply]
I am trying to see if the compiling time would be different. However, after tried your method, it seems like there is still not much change in te proof time. I willl attack the input json I used for testing.

**Reply by 2pir2** [reply]
[inputs.json](https://github.com/user-attachments/files/17300272/inputs.json)

**Reply by 2pir2** [reply]
This one is the large one with large number

**Reply by 2pir2** [reply]
[inputs.json](https://github.com/user-attachments/files/17300294/inputs.json)

**Comment by 2pir2** [comment]
This one is the smaller one. Let me know if there is anything else you want me to provide.

**Comment by ivokub** [comment]
I see it now - here your circuits are exactly the same, you're only changing the input witness. Circuit compile speed doesn't depend on the input. Neither does the proving time unless there is some non-deterministic computation using hints.

NB! All arithmetic in circuit is done modulo the native field. This means that most probably the large input is cut to fit into the native field, as the native field is approx 254 bits.

So the behaviour is expected.

**Reply by 2pir2** [reply]
Thank you so much! I have founded that it is smaller. I would probability try slicing to get the correct hashing.
